### PR TITLE
HIVE-25758: OOM due to recursive application of CBO rules

### PR DESCRIPTION
**Pull request by asolimando**
### What changes were proposed in this pull request?

A notion of "safe" predicate for transitive inference is being introduced. A predicate is unsafe if it can lead to a loop in transitive inference across join operands (due to not being able to simplify redundant predicates). At the moment we only exclude predicates with a disjunction (at any level of nesting).

By default the filtering of unsafe predicate does not happen, which is only triggered if the option `hive.optimize.join.transitive.predicates.conservative=true` is set.

### Why are the changes needed?

Some queries go in OOM (those qtest have been added here).

### Does this PR introduce _any_ user-facing change?

No (the default behaviour is unchanged).

### How was this patch tested?

`mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile="cbo_join_transitive_pred_loop_1.q,cbo_join_transitive_pred_loop_2.q,cbo_join_transitive_pred_loop_3.q,cbo_join_transitive_pred_loop_4.q" -Dtest.output.overwrite -pl itests/qtest -Pitests`

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 1]
```diff
@@ -74,6 +74,8 @@
   public static final HiveJoinPushTransitivePredicatesRule INSTANCE_ANTIJOIN =
           new HiveJoinPushTransitivePredicatesRule(HiveAntiJoin.class, HiveRelFactories.HIVE_FILTER_FACTORY);
 
+  private static final UnsafeOperatorsFinder UNSAFE_OPERATORS_FINDER = new UnsafeOperatorsFinder(true);
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:77]: I think we may end-up calling this visitor from multiple threads so I would either remove static or make the class immutable.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ public void onMatch(RelOptRuleCall call) {
     private final RexBuilder rexBuilder;
     private boolean modified;
 
-    private RexInBetweenExpander(RexBuilder rexBuilder) {
+    RexInBetweenExpander(RexBuilder rexBuilder) {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java:141]: I don't see why this change is necessary. Can you explain?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 3]
```diff
@@ -144,27 +146,70 @@ public void onMatch(RelOptRuleCall call) {
...
 
-    return typeSafeRex;
+    @Override
+    public Void visitCall(RexCall call) {
+      switch (call.getKind()) {
+      case OR:
+        if (inNegation) {
+          return super.visitCall(call);
+        } else {
+          throw Util.FoundOne.NULL;
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:185]: You could possibly avoid using the exception for control flow by simply setting a variable here:
```suggestion
          this.isSafe = false;
          return null;
```

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java (hunk 2 above)
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java:141]: TLDR: Sorry, that's just a leftover from a previous attempt.

Long story:
At some point I attempted to expand `IN` to check the impact on simplification, but there were more side-effects than benefits, I have dropped that part but forgot to put the original visibility back.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 4]
```diff
@@ -144,27 +146,70 @@ public void onMatch(RelOptRuleCall call) {
...
 
     // We need to filter i) those that have been pushed already as stored in the join,
-    // and ii) those that were already in the subtree rooted at child
-    ImmutableList<RexNode> toPush = HiveCalciteUtil.getPredsNotPushedAlready(predicatesToExclude,
-            child, valids);
-    return toPush;
+    // ii) those that were already in the subtree rooted at child,
+    // iii) predicates that are not safe for transitive inference.
+    //
+    // There is no formal definition of safety for predicate inference, only an empirical one.
```
[the remark this answers is not in the document]
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:152]: The change does not reinforce the dependency on the registry. If you check the call, we are filtering candidate predicates coming from `HiveRelMdPredicates`, which is not linked to the registry at all. The predicates in the registry are used to add another filtering step (to remove "duplicates"), but the two steps are totally independent.

This said, I agree that keeping a global state is a problem, and should be avoided. The current implementation is also faulty under some respects.

The motivation for storing predicates in the registry is given in [HIVE-12478](https://issues.apache.org/jira/browse/HIVE-12478?focusedCommentId=15047639&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15047639).

Since predicates can be modified (pushed past project/setOp/joins/aggregate, they can be simplified or merged), we can try to push them transitively as it happens in this OOM cases.

But, as I was saying, the current implementation is faulty, since even the registry is not enough, because most of the other rules they don't know about it, they will alter the predicates, but they won't register them into the registry. Making all those rules aware of the registry would increase the coupling, hence not an option IMO.

The registry main use was to prevent firing the same rule against the same rel multiple times, but even this use is not complete, since rels are modified along the way, and we don't update the registry accordingly. For instance, a join condition is updated, the new join has a fresh rel-id, the registry will try to re-apply `HiveJoinPushTransitivePredicatesRule`, and if the predicates were modified too, this will also escape the duplicate check based on the `registry.getPushedPredicates(join)`, and we can have a loop).

On the other hand, I don't see how we can prevent such loops without storing a global state. But for this global state to work, each and every rule must be aware of it, which is totally not doable IMO, which is also the line in Calcite.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java (hunk 3 above)
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:185]: I know it's an anti-pattern but it's anywhere in the codebase (even in Calcite IIRC I have seen it), but it's better not to add another instance of it.

**Comment by asolimando** [comment]
@zabetak, can I ask you to have another look when you have time?

**Comment by asolimando** [comment]
Gentle reminder for the review @zabetak :)

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:149]: By additional dependency I meant the coupling with `RexSimplify`; the rule does (or doesn't do) something based on what the simplifier can handle. Anyways my comment was not a blocker, but rather a remark to try to address this in a more general fashion in a follow-up. For the moment, I am marking this thread as resolved.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 5]
```diff
@@ -145,19 +138,25 @@ private ImmutableList<RexNode> getValidPreds(RelNode child, Set<String> predicat
...
     // An unsafe predicate in this context is one that when pushed across join operands, can lead
     // to redundant predicates that cannot be simplified (by means of predicates merging with other existing ones).
     // This situation can lead to an OOM for cases where lack of simplification allows inferring new predicates
-    // (from LHS to RHS) recursively, predicates which are redundant, but that RexSimplify cannot handle.
+    // (from LHS to RHS and vice-versa) recursively, predicates which are redundant, but that RexSimplify cannot handle.
     // This notion can be relaxed as soon as RexSimplify gets more powerful, and it can handle such cases.
-    List<RexNode> toPush = HiveCalciteUtil.getPredsNotPushedAlready(predicatesToExclude, child, valids).stream()
-        .filter(unsafeOperatorsFinder::isSafe)
-        .collect(Collectors.toList());
+    if (HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_JOIN_PUSH_TRANSITIVE_PREDICATES_CONSERVATIVE)) {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:155]: Good idea to add a configuration flag for this change.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 6]
```diff
@@ -65,21 +66,15 @@
...
-  public static final HiveJoinPushTransitivePredicatesRule INSTANCE_ANTIJOIN =
-          new HiveJoinPushTransitivePredicatesRule(HiveAntiJoin.class, HiveRelFactories.HIVE_FILTER_FACTORY);
-
+  private final HiveConf conf;
   private final FilterFactory filterFactory;
+  private final UnsafeOperatorsFinder unsafeOperatorsFinder = new UnsafeOperatorsFinder();
 
   public HiveJoinPushTransitivePredicatesRule(Class<? extends Join> clazz,
-      FilterFactory filterFactory) {
+      FilterFactory filterFactory, HiveConf conf) {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:74]: I think the rule is only meant to work with `HiveFilter` so I guess it is safe to remove this parameterization from the constructor. I wouldn't consider CBO rules as public API that should be consumed by clients.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 7]
```diff
@@ -65,21 +66,15 @@
...
-  public static final HiveJoinPushTransitivePredicatesRule INSTANCE_JOIN =
-          new HiveJoinPushTransitivePredicatesRule(HiveJoin.class, HiveRelFactories.HIVE_FILTER_FACTORY);
-
-  public static final HiveJoinPushTransitivePredicatesRule INSTANCE_SEMIJOIN =
-          new HiveJoinPushTransitivePredicatesRule(HiveSemiJoin.class, HiveRelFactories.HIVE_FILTER_FACTORY);
-
-  public static final HiveJoinPushTransitivePredicatesRule INSTANCE_ANTIJOIN =
-          new HiveJoinPushTransitivePredicatesRule(HiveAntiJoin.class, HiveRelFactories.HIVE_FILTER_FACTORY);
-
+  private final HiveConf conf;
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:69]: Instead of keeping the whole config it may be better to just keep the boolean variable that we need. Reduces the likelihood of a memory leak due to configuration objects.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 8]
```diff
@@ -143,28 +138,82 @@ private ImmutableList<RexNode> getValidPreds(RelOptCluster cluster, RelNode chil
...
+  /**
+   * Finds unsafe operators in an expression (at any level of nesting).
+   * At the moment, the only unsafe operator is OR.
+   *
+   * Example 1: OR(=($0, $1), IS NOT NULL($2))):INTEGER (OR in the top-level expression)
+   * Example 2: NOT(AND(=($0, $1), IS NOT NULL($2))
+   *   this is equivalent to OR((<>($0, $1), IS NULL($2))
+   * Example 3: AND(OR(=($0, $1), IS NOT NULL($2)))) (OR in inner expression)
+   */
+  private static class UnsafeOperatorsFinder extends RexVisitorImpl<Void> {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:175]: If I understand well this class tries to find disjunctions inside an expression. Instead of using the broader term "unsafe" I would focus and rename usages/documentation to be more explicit. This would make the code easier to understand and   plan changes due to this self-explained.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java [hunk 9]
```diff
@@ -143,28 +138,82 @@ private ImmutableList<RexNode> getValidPreds(RelOptCluster cluster, RelNode chil
...
+      case NOT:
+        inNegation = !inNegation;
+        return super.visitCall(call);
+      default:
+        return super.visitCall(call);
+      }
     }
 
-    return typeSafeRex;
+    boolean isSafe(RexNode node) {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:209]: Consider making this method a utility by moving it in `HiveCalciteUtil` e.g., `HiveCalciteUtil#containsDisjunction`. Also maybe it would be safer to make the class disposable, i.e., instantiate only inside the method and then discard. We are going to spend some extra resources for instantiation but I guess the overhead is negligible.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java (hunk 8 above)
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:175]: If in the future we find out that we want to cover more use-cases we can revise the naming.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 10]
```diff
@@ -2547,6 +2547,9 @@ public static enum ConfVars {
         "If this config is true only pushed down filters remain in the operator tree, \n" +
         "and the original filter is removed. If this config is false, the original filter \n" +
         "is also left in the operator tree at the original place."),
+    HIVE_JOIN_PUSH_TRANSITIVE_PREDICATES_CONSERVATIVE("hive.optimize.join.transitive.predicates.conservative",
+        false, "Whether to avoid pushing predicates that are hard to simplify. \n"
```
Comment (zabetak) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2551]: I am not sure if we want to keep this so general or focus on what actually happens, which is forbidding disjunction pushdown (e.g., `hive.optimize.join.disjunctive.transitive.predicates.pushdown`).

Regarding the default value maybe it would be better to disallow the optimization by default. From my perspective, correctness is more important than performance so having HS2 crashing with OOM is more serious than having perf regression in some queries. 

Both of the comments above are rather personal preferences so I am not pushing hard for making these changes. The final decision is up to you :)

**Review by zabetak — COMMENTED** [review]
Overall LGTM, just left a few minor comments. I will merge this back when we close the pending discussions and get a green run!

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java (hunk 7 above)
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java:69]: This is a great idea, I am doing that right away.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 10 above)
Comment (asolimando) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2551]: Since we are focusing on the disjunctive predicates (as per another comment), let's opt for `hive.optimize.join.disjunctive.transitive.predicates.pushdown`.

As discussed offline, let's leave the property disabled by default, because it's hard to judge how much benefit/harm there will be for Hive in the wild, we can revisit the default value later on if we have more evidence that this will help.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java [hunk 11]
```diff
@@ -1214,6 +1214,58 @@ public FixNullabilityShuttle(RexBuilder rexBuilder,
...
 
+  /**
+   * Find disjunction (OR) in an expression (at any level of nesting).
+   *
+   * Example 1: OR(=($0, $1), IS NOT NULL($2))):INTEGER (OR in the top-level expression)
+   * Example 2: NOT(AND(=($0, $1), IS NOT NULL($2))
+   *   this is equivalent to OR((<>($0, $1), IS NULL($2))
+   * Example 3: AND(OR(=($0, $1), IS NOT NULL($2)))) (OR in inner expression)
+   */
+  public static class DisjunctivePredicatesFinder extends RexVisitorImpl<Void> {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java:1225]: How about making this class private and expose only a method? `public static hasDisjunction`?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 12]
```diff
@@ -2547,6 +2547,9 @@ public static enum ConfVars {
         "If this config is true only pushed down filters remain in the operator tree, \n" +
         "and the original filter is removed. If this config is false, the original filter \n" +
         "is also left in the operator tree at the original place."),
+    HIVE_JOIN_DISJ_TRANSITIVE_PREDICATES_PUSHDOWN("hive.optimize.join.disjunctive.transitive.predicates.pushdown",
+        true, "Whether to transitively infer disjunctive predicates across joins. \n"
+            + "Disjunctive predicates can lead to OOM in transitive inference."),
```
Comment (zabetak) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:2552]: Suggestion: Disjunctive predicates are hard to simplify and pushing them down may in some cases lead to infinite rule matching causing stackoverflow and OOM errors.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java [hunk 13]
```diff
@@ -1214,6 +1214,58 @@ public FixNullabilityShuttle(RexBuilder rexBuilder,
...
   }
 
+  /**
+   * Find disjunction (OR) in an expression (at any level of nesting).
+   *
+   * Example 1: OR(=($0, $1), IS NOT NULL($2))):INTEGER (OR in the top-level expression)
+   * Example 2: NOT(AND(=($0, $1), IS NOT NULL($2))
+   *   this is equivalent to OR((<>($0, $1), IS NULL($2))
+   * Example 3: AND(OR(=($0, $1), IS NOT NULL($2)))) (OR in inner expression)
+   */
```
[the remark this answers is not in the document]
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java:1224]: I get no errors while compiling (without `-Dmaven.javadoc.skip` of course).

Is there anything specific you are referring to?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java [hunk 14]
```diff
@@ -1257,15 +1249,26 @@ public Void visitCall(RexCall call) {
...
+   * Example 1: OR(=($0, $1), IS NOT NULL($2))):INTEGER (OR in the top-level expression)
+   * Example 2: NOT(AND(=($0, $1), IS NOT NULL($2))
+   *   this is equivalent to OR((<>($0, $1), IS NULL($2))
+   * Example 3: AND(OR(=($0, $1), IS NOT NULL($2)))) (OR in inner expression)
+   *
+   * @param node the expression where to look for disjunctions.
+   * @return true if the given expressions contains a disjunction, false otherwise.
+   */
+  public static boolean hasDisjuction(RexNode node) {
+    return new DisjunctivePredicatesFinder().hasDisjunction(node);
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java:1269]: nit to get rid of the `DisjunctivePredicateFinder#hasDisjunction` method:
```
DisjunctivePredicateFinder finder = new DisjunctivePredicateFinder();
node.accept(finder);
return finder.hasDisjunction;
```

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java [hunk 15]
```diff
@@ -1257,15 +1249,26 @@ public Void visitCall(RexCall call) {
...
-      inNegation = false;
-      hasDisjunction = false;
-
       node.accept(this);
       return hasDisjunction;
     }
   }
 
+  /**
+   * Find disjunction (OR) in an expression (at any level of nesting).
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java:1258]: `Find` implies that we are going to return a disjunctive expression. Suggestion:
`Returns whether the expression has disjunctions (OR) at any level of nesting.`

**Comment by asolimando** [comment]
Tests are finally green, is there anything else that is needed @zabetak?
